### PR TITLE
Implement cast

### PR DESCRIPTION
**Pull request by mei1127**
@BruceDai @huningxin @fdwr PTAL,thanks!

### CODE REVIEW

File: test/cast_test.js [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+describe('test cast', function() {
+  function testCast(input, type, expected) {
+    const tensorInput = new Tensor(input.shape, input.data);
+    const outputTensor = cast(tensorInput, type);
+    utils.checkShape(outputTensor, expected.shape);
+    utils.checkValue(outputTensor, expected.data);
+  }
+
+  it('cast float32', function() {
```
Comment (BruceDai) [inline test/cast_test.js:15]: Please rename test name as "cast float64 to float32", and same to others tests.

File: src/cast.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    case 'float32':
+      outputArray = new Float32Array(input.data);
+      break;
+    case 'int32':
+      outputArray = new Int32Array(input.data);
+      break;
+    case 'uint32':
+      outputArray = new Uint32Array(input.data);
+      break;
+    case 'int64':
```
Comment (BruceDai) [inline src/cast.js:23]: Please fix this error "TypeError: Cannot convert a BigInt value to a number" of CI.

File: src/cast.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    case 'int8':
+      outputArray = new Int8Array(input.data);
+      break;
+    case 'uint8':
+      outputArray = new Uint8Array(input.data);
+      break;
+    case 'float16':
+      // todo
+      throw new Error('Unsupported output type: float16' );
+    // case 'uint64':
```
Comment (BruceDai) [inline src/cast.js:35]: There're two commented **case** block for **uint64** type, please update it.

File: test/utils.js [hunk 4]
```diff
@@ -27,11 +27,17 @@ function getBitwise(value) {
...
-  distance = distance >= 0 ? distance : -distance;
-  return assert.isTrue(distance <= nulp, message);
+  if (typeof(a) == 'number') {
+    const aBitwise = getBitwise(a);
+    const bBitwise = getBitwise(b);
+    let distance = aBitwise - bBitwise;
+    distance = distance >= 0 ? distance : -distance;
+    return assert.isTrue(distance <= nulp, message);
+  } else {
+    let distance = a-b;
```
Comment (fdwr) [inline test/utils.js:37]: [nit] `a - b` consistent spacing with line 33.

Comment (fdwr) [reply test/utils.js:37]: (interesting - so in Javascript, int64 is considered not a "number" 🙃)

File: test/cast_test.js [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+    const input = {
+      shape: [5],
+      data: [
+        -0.25, 0.25, 3.21, 1234, -1234,
+      ],
+    };
+    const expected = {
+      shape: [5],
+      data: [
+        0, 0, 3, 210, 46,
```
[the remark this answers is not in the document]
Comment (fdwr) [reply test/cast_test.js:55]: Related: https://github.com/webmachinelearning/webnn/issues/489

**Review by fdwr — COMMENTED** [review]
Some bounds considerations and test case addition requests, but otherwise LGTM.

File: test/cast_test.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (mei1127) [reply test/cast_test.js:55]: Okay, thank! I'll convert examples outside the type range into within the range.

File: test/cast_test.js [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+    const outputTensor = cast(tensorInput, type);
+    utils.checkShape(outputTensor, expected.shape);
+    utils.checkValue(outputTensor, expected.data);
+  }
+
+  it('cast float64 to int8', function() {
+    const input = {
+      shape: [5],
+      data: [
+        -0.25, 0.25, 3.21, 1234, -1234,
```
[the remark this answers is not in the document]
Comment (mei1127) [reply test/cast_test.js:33]: ok

File: test/cast_test.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (huningxin) [reply test/cast_test.js:55]: +1, let's focus on the defined behavior at this stage.

File: test/utils.js (hunk 4 above)
Comment (mei1127) [reply test/utils.js:37]: The integer of type INT64 exceeds the range of JavaScript's Number type and belongs to the integer of BigInteger type. The JavaScript exception "can't convert BigInt to number" occurs when an arithmetic operation involves a mix of BigInt and Number values. FYI: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/Cant_convert_BigInt_to_number?retiredLocale=pl

File: src/cast.js [hunk 7]
```diff
@@ -12,16 +12,16 @@ export function cast(input, type) {
...
-      outputArray = new Uint8Array(input.data);
+      outputArray = new Uint8Array(Array.from(input.data, (num) => (Math.round(num))));
       break;
     case 'int32':
-      outputArray = new Int32Array(input.data);
+      outputArray = new Int32Array(Array.from(input.data, (num) => (Math.round(num))));
       break;
     case 'uint32':
-      outputArray = new Uint32Array(input.data);
+      outputArray = new Uint32Array(Array.from(input.data, (num) => (Math.round(num))));
```
Comment (fdwr) [inline src/cast.js:24]: Truncation toward zero is the expected casting behavior for float to int, not rounding to nearest. That's consistent with existing ML libraries I know of. e.g. `float32` 3.7 -> `uint32` 3. Isn't the default behavior of `Uint32Array` already the expected truncation toward zero behavior?

Comment (mei1127) [reply src/cast.js]: I'm sorry for the misunderstanding. I've revised the code :)

**Review by fdwr — APPROVED** [review]
👍 TY Mei.

File: src/cast.js [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+      outputArray = new BigInt64Array(Array.from(input.data, (num) => BigInt(Math.trunc(num))));
+      break;
+    case 'float32':
+      outputArray = new Float32Array(input.data);
+      break;
+    case 'float64':
+      outputArray = new Float64Array(input.data);
+      break;
+    case 'float16':
+      // todo
```
Comment (huningxin) [inline src/cast.js:36]: Please file an issue tracking this task and put the issue link into the TODO, for example
```
// TODO(issue_link): Implement float16 output type.
```

File: src/cast.js [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+      throw new Error('Unsupported output type: float16' );
+    case 'uint64':
+      // todo
```
Comment (huningxin) [inline src/cast.js:39]: ditto

**Review by fdwr — APPROVED** [review]

**Review by huningxin — APPROVED** [review]

**Pull request merged**